### PR TITLE
Limit tech admin contact operations via IS config

### LINKED ISSUE #6323: IS config for limiting who can be a admin/tech contact for entities
### Summary

References #4930

Like suggested by Johan on the issue, instead of privacy flags there should be IdentityServer configuration option which restricts an user from setting another collaborator as a administrative/technical contact.

With this, operations that could abuse the privacy of other users are restricted.  In addition to that, it removes the necessity of privacy flags on the `Collaborator` entity.

### Current Situation

Administrative/Technical contacts are already implemented and they replace the previous `ContactInfo`, but with the current implementation any collaborator would be able to set another collaborator as one of these contacts. 

### Why do we need this? Who uses it, and when?

A IS config option which restricts whom an user can set as a contact to only himself.

### Proposed Implementation

Add an extra field to `IsConfiguration` which disallows the user to set others as the administrative/technical contact.

### Contributing

- [ ] I can help by doing more research.
- [X] I can help by implementing the feature after the proposal above is approved.
- [X] I can help by testing the feature before it's released.

### Code of Conduct

- [X] I agree to follow TTN's [Community Code of Conduct](https://github.com/TheThingsNetwork/Manifesto/blob/master/CodeOfConduct.md).

### PR DESCRIPTION
**Pull request by nicholaspcr**
#### Summary

Closes #6323

#### Changes

- Add an contact info restriction error
- Add `restrictions` field to IS config, alongside the contact info restriction
- Update entities registries to validate contact if config is enabled

#### Testing

Unit tests and manual testing

##### Regressions

This is a new restriction on any flow that sets the admin/tech contacts which aren't done through the console. So I can't imagine a regression and most likely CLI users aren't setting those contacts since the old way still works (even with the deprecated warning).

#### Notes for Reviewers

Was going to do a testcase just for restrictions for each registry test file but it felt like overkill, let me know if the smaller tests found currently on the PR are appropriate enough.

If there are any lint errors on CI, I'll run the tests again after the format merge of the separate repository. Made the lint changes into a different PR just to remove it from this PR.

#### Checklist

- [x] Scope: The referenced issue is addressed, there are no unrelated changes.
- [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: api/identityserver.proto [hunk 1]
```diff
@@ -125,6 +125,19 @@ message IsConfiguration {
   reserved 10; reserved "application_limits";
   reserved 11; reserved "organization_limits";
   reserved 12; reserved "user_limits";
+
+  // Restrictions contains constraints that are applied to the IS's business logic.
+  message Restrictions {
+    // Entity contains restrictions that are applied to entities.
+    message Entity {
+      // restrict_contacts forbids standard users from setting others as the administrative or technical contacts.
+      google.protobuf.BoolValue restrict_contacts = 1;
```
Comment (johanstokking) [inline api/identityserver.proto]: I would make this more explicit. The comment is clear but the field name is not really. Maybe `allow_set_others_as_contact` ?

Also nit; no need for Go style comments here, 
```suggestion
      // Allow anyone to set others as the administrative or technical contacts.
      google.protobuf.BoolValue allow_set_others_as_contact = 1;
```

Comment (nicholaspcr) [reply api/identityserver.proto]: Indeed the name could be better but I do disagree with the name suggestion slightly. Since this is within the restriction config, it makes more sense in my eyes to have variable which are capable of restriction rather than allowing it all operations.

TLDR: probably more appropriate to have something like `disallow_set_other_as_contact`

Will do the changes in a bit, regarding the `nit`, I know its rather insignificant but its mostly something for when someone wants to check the GoDoc of the generated code

File: pkg/identityserver/config.go [hunk 2]
```diff
@@ -115,6 +115,11 @@ type Config struct {
 		TenantID string         `name:"tenant-id" description:"Tenant ID in the host NetID"`
 	} `name:"network"`
 	TelemetryQueue telemetry.TaskQueue `name:"-"`
+	Restrictions   struct {
+		Entity struct {
+			DisallowSetOtherAsContact bool `name:"disallow-set-other-as-contact" description:"Restrict users from setting others as the administrative or technical contact of any entity"`
```
Comment (adriansmares) [inline pkg/identityserver/config.go]: How about `CollaboratorRights.SetOthersAsContacts` ? We already have `UserRights` and `AdminRights`, and `SetOthersAsContacts` (default `false`) would disallow collaborators from setting the contacts to other values other than their own identity.

File: pkg/identityserver/application_registry_test.go [hunk 3]
```diff
@@ -187,6 +187,32 @@ func TestApplicationsCRUD(t *testing.T) {
 			a.So(updated.Name, should.Equal, "Updated Name")
 		}
 
+		t.Run("contact_info_restriction", func(t *testing.T) { // nolint:paralleltest
+			a, ctx := test.New(t)
+			is.config.Restrictions.Entity.DisallowSetOtherAsContact = true
+			t.Cleanup(func() { is.config.Restrictions.Entity.DisallowSetOtherAsContact = false })
```
Comment (adriansmares) [inline pkg/identityserver/application_registry_test.go]: You should set the old value, not `false`.

File: pkg/identityserver/store/errors.go [hunk 4]
```diff
@@ -124,4 +124,8 @@ var (
 		"application_dev_eui_limit_reached",
 		"application issued DevEUI limit ({dev_eui_limit}) reached",
 	)
+
+	ErrContactInfoRestricted = errors.DefinePermissionDenied(
```
Comment (adriansmares) [inline pkg/identityserver/store/errors.go:128]: Maybe the description should be `contact information can only reference to the caller` ?

**Review by adriansmares — COMMENTED** [review]
Implementation LGTM in general. I have some naming concerns only.

File: pkg/identityserver/config.go (hunk 2 above)
Comment (nicholaspcr) [reply pkg/identityserver/config.go]: I did think about doing something similar to `AdminRights` but for Colalborators like you mentioned but the way I'm tackling this mentally is more in regards to how classify these sort of operations. 

I was going to write that I'm seeing this more as a set of operations restricted by deployment configurations than `Rights` but it should be able to be bypassed by admins (which currently it doesn't) so indeed your suggestion probably makes more sense

File: api/identityserver.proto [hunk 5]
```diff
@@ -125,6 +125,11 @@ message IsConfiguration {
   reserved 10; reserved "application_limits";
   reserved 11; reserved "organization_limits";
   reserved 12; reserved "user_limits";
+
+  message CollaboratorRights {
+    google.protobuf.BoolValue set_other_as_contact = 1;
```
Comment (adriansmares) [inline api/identityserver.proto]: `set_others_as_contacts`.

File: pkg/identityserver/config.go [hunk 6]
```diff
@@ -83,6 +83,9 @@ type Config struct {
 	AdminRights struct {
 		All bool `name:"all" description:"Grant all rights to admins, including _KEYS and _ALL"`
 	} `name:"admin-rights"`
+	CollaboratorRights struct {
+		SetOthersAsContacts bool `name:"set-others-as-contacts" description:"Allow users to set other users as their contact"`
```
Comment (adriansmares) [inline pkg/identityserver/config.go]: `as entity contacts`.

File: pkg/identityserver/gateway_registry_test.go [hunk 7]
```diff
@@ -207,6 +207,56 @@ func TestGatewaysCRUD(t *testing.T) {
 			a.So(updated.Name, should.Equal, "Updated Name")
 		}
 
+		t.Run("contact_info_restrictions", func(t *testing.T) { // nolint:paralleltest
```
Comment (adriansmares) [inline pkg/identityserver/gateway_registry_test.go]: Spaces are automatically converted to underlines, so you can just leave the spaces in the naming, or use something like `ContactInfoRestrictions`. We don't have a consistent style in the codebase.

File: pkg/identityserver/store/errors.go (hunk 4 above)
Comment (adriansmares) [reply pkg/identityserver/store/errors.go:128]: Apparently it should be `contact information can only reference the caller`, apologies.

**Review by adriansmares — APPROVED** [review]

File: pkg/identityserver/gateway_registry_test.go (hunk 7 above)
Comment (nicholaspcr) [reply pkg/identityserver/gateway_registry_test.go]: Was using the underscore to make it easier to skip the whole test name under my text editor, since I don't want to create a third way of doing things I changed to the first suggestion. Thanks for catching that.

File: CHANGELOG.md [hunk 8]
```diff
@@ -15,6 +15,7 @@ For details about compatibility between different releases, see the **Commitment
 - RPCs and CLI command to delete a batch of end devices within an application.
   - Check `ttn-lw-cli end-devices batch-delete` for more details.
 - Add `UserInput` component to the Console to handle user id input fields by implementing an autosuggest.
+- The Identity Server configuration has a new optional configurable restriction regarding adminstrative and technical contacts of entities. This limits the action of an user or organization to set these contacts only to themselves, its enabled by default but it is possible to disable it by setting `is.collaborator-rights.set-others-as-contacts` as true.
```
Comment (adriansmares) [inline CHANGELOG.md]: I think we should default to `true` in order to avoid breaking changes.

Comment (nicholaspcr) [reply CHANGELOG.md]: yeah, I agree
